modules/renter/contractor/contracts.go

```diff
@@ -83,6 +84,119 @@ func (c *Contractor) contractEndHeight() types.BlockHeight {
 	return endHeight
 }
 
+// managedMarkContractsUtility checks every active contract in the contractor and
+// figures out whether the contract is useful for uploading, and whehter the
+// contract should be renewed.
+func (c *Contractor) managedMarkContractsUtility() {
+	// Pull a new set of hosts from the hostdb that could be used as a new set
+	// to match the allowance. The lowest scoring host of these new hosts will
+	// be used as a baseline for determining whether our existing contracts are
+	// worthwhile.
+	c.mu.RLock()
+	hostCount := int(c.allowance.Hosts)
+	c.mu.RUnlock()
+	if hostCount <= 0 {
+		return
+	}
+	hosts := c.hdb.RandomHosts(hostCount+minScoreHostBuffer, nil)
+	if len(hosts) <= 0 {
+		return
+	}
+
+	// Find the lowest score of this batch of hosts.
+	lowestScore := c.hdb.ScoreBreakdown(hosts[0]).Score
+	for i := 1; i < len(hosts); i++ {
+		score := c.hdb.ScoreBreakdown(hosts[i]).Score
+		if score.Cmp(lowestScore) < 0 {
+			lowestScore = score
+		}
+
+	}
+	// Set the minimum acceptable score to a factor of the lowest score.
+	minScore := lowestScore.Div(scoreLeeway)
+
+	// Pull together the set of contracts.
+	c.mu.RLock()
+	contracts := make([]modules.RenterContract, 0, len(c.contracts))
+	for _, contract := range c.contracts {
+		contracts = append(contracts, contract)
+	}
+	c.mu.RUnlock()
+
+	// Go through and figure out if the utility fields need to be changed.
+	for i := 0; i < len(contracts); i++ {
+		// Start the contract in good standing.
+		contracts[i].GoodForUpload = true
+		contracts[i].GoodForRenew = true
+
+		host, exists := c.hdb.Host(contracts[i].HostPublicKey)
+		// Contract has no utility if the host is not in the database.
+		if !exists {
+			contracts[i].GoodForUpload = false
+			contracts[i].GoodForRenew = false
+			continue
+		}
+		// Contract has no utility if the score is poor.
+		if c.hdb.ScoreBreakdown(host).Score.Cmp(minScore) < 0 {
+			contracts[i].GoodForUpload = false
+			contracts[i].GoodForRenew = false
+			continue
+		}
+		// Contract has no utility if the host is offline.
+		c.mu.Lock()
+		offline := c.isOffline(contracts[i].ID)
+		c.mu.Unlock()
```

Comment:
convention mandates that any call to an unexported function elsewhere in the module requires a writelock.